### PR TITLE
Webcam draggable refactoring

### LINKED ISSUE #7598: Follow-up to #7589 Draggable webcams
#7589 

- [ ] On Firefox webcams do not fill up the screen when the presentation area is minimized. This is not an issue on Chrome
- [ ] Multiple webcams do not utilize the area well (compared to <4 webcams) when the presentation area is minimized
Screenshot from 2019-06-10 16-01-11

- [ ] Single click on webcams can cause the draggable webcams to overlay the presentation
ClickOnWebcam

- [ ] Improve on .5sec delay:
> Can't the webcam pos/reference point calls be tethered to component state changes, then? I see that they're enveloped on async setTimeouts on almost every call when there's a change event of some sort (focus, resize). I worry it might be flaky on low end computers anyways (timeouts always are).

https://github.com/bigbluebutton/bigbluebutton/pull/7589#discussion_r292160105

### PR DESCRIPTION
**Pull request by vitormateusalmeida**
Closes #7598 

swap-multiples
drag-multiples
overflow
single-to-multiple
swap-single
userlist-toggle

### CODE REVIEW

File: bigbluebutton-html5/imports/ui/components/app/component.jsx [hunk 1]
```diff
@@ -266,6 +266,9 @@ class App extends Component {
         className={styles.media}
         aria-label={intl.formatMessage(intlMessages.mediaLabel)}
         aria-hidden={this.shouldAriaHide()}
+        style={{
```
Comment (Tainan404) [inline bigbluebutton-html5/imports/ui/components/app/component.jsx]: It could add it to css class related to this component.

**Comment by Tainan404** [comment]
I don't what element should have preference in z-index, but currently the webcam overlaps the whiteboard tool-bar and the user-list dropdown.
cc @antobinary.

Peek 2019-07-10 16-56

Peek 2019-07-10 16-57

File: bigbluebutton-html5/imports/ui/components/media/webcam-draggable-overlay/component.jsx [hunk 2]
```diff
@@ -1,596 +1,305 @@
 import React, { Component, Fragment } from 'react';
-import PropTypes from 'prop-types';
+import Draggable from 'react-draggable';
 import cx from 'classnames';
-import VideoProviderContainer from '/imports/ui/components/video-provider/container';
-import _ from 'lodash';
+import * as _ from 'lodash';
```
Comment (Tainan404) [inline bigbluebutton-html5/imports/ui/components/media/webcam-draggable-overlay/component.jsx]:
```suggestion
import _ from 'lodash';
```

File: bigbluebutton-html5/imports/ui/components/video-provider/video-list/video-list-item/component.jsx [hunk 3]
```diff
@@ -198,7 +211,7 @@ class VideoListItem extends Component {
   }
 }
 
-export default injectIntl(VideoListItem);
+export default injectIntl(withConsumer(VideoListItem));
```
Comment (Tainan404) [inline bigbluebutton-html5/imports/ui/components/video-provider/video-list/video-list-item/component.jsx]: Could you put a meaningful name for the consumer, that's very generic.

**Comment by prlanzarin** [comment]
Spent quite some time *testing* this, seems very good. Cases that were broken before don't seem to break now. React updates seem fairly contained as well, didn't notice any CPU hogs of any sort. Neither console exceptions.

I tested this in Firefox and Chrome for desktop and mobile. **Could use some testing on Edgium and Safari (macOS and iOS, both) before merging** (or someone's assurance on that front).

Didn't read the code thoroughly to have anything meaningful to say about it (too many changes for my weak react-fu); so this is just sort of a manual testing review.

Did notice a small quirk on presentation minimize -> maximize: with a single webcam positioned **off** the top/bottom anchor points, there's an odd transition step on pres. maximize where the cam will be shown in the some random presentation edge before transitioning to its original position (see .gif). It's small, but noticeable. Might as well be painful to solve properly, so it's not a deal breaker.

dae-ze

**Pull request merged**